Add option to ignore network errors when attempting to log

kschat opened the pull request:
Adds ability to ignore network errors when attempting to send logs so logging can continue, even if a single log failed to send. For instance, we were seeing all our logs drop off because a single request to our logging service received an `ECONNRESET`.

arb commented on lib/index.js:
```diff
@@ -68,7 +69,11 @@ class GoodHttp extends Stream.Writable {
+        Wreck.request('post', this._endpoint, wreckOptions, (err, payload) => {
+
+            err = this._settings.ignoreRequestErrors ? null : err;
+            return callback && callback(err, payload);
```
I think you should move this logic to line 50 since there already is logic there for dealing with error type conditions. Additionally you can just do `callback(err, payload);` no need for the extra `callback &` thing.

arb commented:
Now that I think about it... I'm not sure if this is a good approach... we ignore errors and clear the `data` array. So we've essentially lost all logging information that might have occurred from one log to the next.

kschat commented:
Would you prefer if there was retry logic? Ultimately I'd just like a way to configure good-http to not be destroyed in the event that one log failed.

arb commented:
How about we split the difference... if there is an error, don't clear the `data` array every time. Instead have like a failure count that once it hits some configurable threshold, clear `data` and reset the threshold. Sort of like a cheap retry?

kschat commented:
Sure, that sounds good. I'll make the necessary changes.

kschat commented:
Updated, let me know what you think.

FYI, looks like Lab updated eslint to version 3 which flagged an issue with an older test. I fixed it in this PR.

arb commented on lib/index.js:
```diff
@@ -12,7 +12,9 @@ const Wreck = require('wreck');
         threshold: 20,
+        errorThreshold: 0,
         schema: 'good-http',
+        ignoreRequestErrors: false,
```
Doesn't `errorThreshold` imply `ignoreRequestErrors`? Can we get by with just `errorThreshold`? If you set `errorThreshold` to anything above 0, doesn't the mean you want to ignore request errors?

kschat replied:
I kept `ignoreRequestErrors` because `errorThreshold` only supports ignoring a finite amount of errors, but not all errors. Is "ignore all errors" a use case you'd like to support? If so, we can still remove `ignoreRequestErrors` by making `errorThreshold: -1` mean "ignore all errors".

arb replied:
You'd set `errorThreshold` to `Infinity` to ignore all errors I would think. Default it to `0` so it works like it does now, `Infinity` means ignore all, and a number means allow that many like per log attempt.

Sound reasonable to you?

kschat replied:
The only problem with that is `Infinity` is not part of the JSON spec so anyone using JSON to configure Good will not be able to use that option. `Infinity` is serialized to `null` so we could workaround it by test for `null` as well.

arb replied:
Good call out. Let's do that then.

arb commented on lib/index.js:
```diff
@@ -44,8 +47,16 @@ class GoodHttp extends Stream.Writable {
         if (this._data.length >= this._settings.threshold) {
             this._sendMessages((err) => {
 
+                if (err && this._failureCount < this._settings.errorThreshold - 1) {
```
Can this be `err && this._failureCount <= this._settings.errorThreshold`?

arb commented on lib/index.js:
```diff
@@ -44,8 +47,16 @@ class GoodHttp extends Stream.Writable {
+                    return callback();
+                }
+
                 // always clear the data so we don't buffer this forever if there is ever a failed POST
```
Remove this comment since it doesn't really make sense any more.

kschat commented:
Removed `ignoreRequestErrors` in favor of the logic discussed above.

arb commented on lib/index.js:
```diff
@@ -12,25 +12,39 @@ const Wreck = require('wreck');
         config = config || {};
+
+        // Infinity is not available in JSON, so assume explicit null means Infinity
+        if (config.errorThreshold === null) {
```
I think you can move this logic down to line 41 rather than doing another `Object.assign`.

kschat replied:
Oh, right. I was thinking `Object.assign` wouldn't overwrite the default with `null`.

arb commented on README.md:
```diff
@@ -18,7 +18,8 @@ Creates a new GoodHttp object where:
 - `config` - configuration object
-	- `[threshold]` - number of events to hold before transmission. Defaults to `20`. Set to `0` to have every event start transmission instantly. It is strongly suggested to have a set threshold to make data transmission more efficient.
+  - `[threshold]` - number of events to hold before transmission. Defaults to `20`. Set to `0` to have every event start transmission instantly. It is strongly suggested to have a set threshold to make data transmission more efficient.
+  - `[errorThreshold]` - number of consecutive failed transmissions allowed (`ECONNRESET`, `ECONNREFUSED`, etc). Defaults to `0`. Failed events will be included in the next transmission until they are successfully logged or the threshold is reached (whichever comes first) at which point they will be cleared. Set to `Infinity` (or `null` if configuration is JSON) to ignore all errors (*note: this  ...
```
Sorry for so much back and forth... but seeing it helps me think about the problem better.

- `0` - means fail on errors (this is what currently happens)
- `> 0` - means allow that many failures before flushing `data` and reporting the error
- `null/Infinity` - means keep trying and never `flush` data

Now that I see the hassle of the `null` `Infinity` thing, let's just use `null`. 

Also, in instance where it's `null` I think line 69 should be something like `callback (this._settings.errorThreshold !== null && err)` or something like that. To me `null` means I don't care about errors and I just want this to keep processing.

What do you think?

kschat commented:
Added suggested changes.

arb commented on test/index.js:
```diff
@@ -244,4 +242,128 @@ describe('GoodHttp', () => {
+        server.listen(0, '127.0.0.1', () => {
+
+            const stream = internals.readStream();
+            const reporter = new GoodHttp(internals.getUri(server), {
```
Is there a way to show that `reporter.data` has stuff in it before "error" happens?

kschat replied:
The assertions in the request `end` handler are testing that the data is being held upon error. I can try to add a more explicit test if you'd like, however.

arb replied:
My point is that you aren't asserting that `data` has stuff in it before the error threshold happens. The default case `data` would clear out after every call, so we need to make sure this change works as expected. So this test isn't really testing what the description is.

Also can you someone assert that the error handler is only called a single time? That's the other key part of this test.

kschat replied:
Right, and the assertions in the request `end` handler are testing that the events sent to the server are including previous failed events, which would only happen if `data` wasn't being cleared out on every call.

From what I can tell there isn't an easy way to test that `reporter._data` contains data before the error event is triggered within an integration test (a unit test would be possible, though).

I can update the test description to read something along the lines of "includes failed logs until errorThreshold is reached" so the assertions are more inline with the description.

kschat commented:
I added some more assertions to the `errorThreshold` test.

arb commented on lib/index.js:
```diff
@@ -12,13 +12,18 @@ const Wreck = require('wreck');
+    host: Os.hostname(),
+    shouldHoldData(errorThreshold, failureCount) {
+
+        return errorThreshold !== null && failureCount < errorThreshold;
```
Sorry I've been slow on this; I went on vacation while we were working on this... so my idea was during initialization, if `errorThreshold` is `null`, we set it internally to `Infinity`, then we can skip this annoying `null` check and actually just get rid of the `shouldHoldData` function completely since then we can just do a numeric comparison between `this._settings.errorThreshold` and `this._failureCount`

kschat replied:
If we set it to `Infinity` internally then we'd still need to do an explicit check for `Infinity` here and on line 61 in order to continue to support this behavior: "Set to `null` to ignore all errors and always clear events".

arb replied:
You can do `<` comparisons directly against `Infinity`. That way you don't need the explicit `null` check.

```js
5 < Infinity true
-1 < Infinity true
```

My goal is to turn `null` into something we don't have to explicitly check for and I think `Infinity` let's us do that.

kschat replied:
The decided behavior when `errorThreshold` is `null` is to "never report errors and never hold events". If we internally set `errorThreshold` to `Infinity` and just do a numeric comparison between `failureCount` and `errorThrshold` then we lose the "never hold events" behavior and it could lead to a memory leak and/or a huge request.

We can internally set `errorThreshold` to `Infinity`, but we will still have to explicitly test for it in `shouldHoldData` (`errorThreshold !== Infinity && failureCount < errorThreshold`), and on line [61](https://github.com/hapijs/good-http/pull/40/files#diff-6d186b954a58d5bb740f73d84fe39073R61).

e.g. if `errorThreshold` is `Infinity` we should **not** hold data.

arb replied:
Make it `-Infinity` then?

kschat commented:
Updated to `-Infinity` implementation
